D96-D98 helper routines ffdb_* (dump writers, units 1081-1085; ffd_aflux_*, ffd_pgf_*, ffd_advecv_* files). Generated by diff -u against the PRISTINE ATM_DRV.f (2026-10-05); anchored after 'end subroutine finalize_atm' (not used by any other patch), so it applies to the pristine file or after any other ATM_DRV_*.patch.

--- ATM_DRV.f.orig
+++ ATM_DRV.f
@@ -1069,6 +1069,348 @@
       end subroutine finalize_atm
 
 
+      logical function ffdb_on()
+!@sum full-fidelity-port instrumentation (D96-D98): true when env
+!@+   FFD_START/FFD_NSTEP select this itime.  Units 1081-1085.
+      use model_com, only : itime
+      implicit none
+      integer, save :: it0=-1, nst=0, init=0
+      integer :: st
+      character(len=64) :: buf
+      if (init.eq.0) then
+        init=1
+        call get_environment_variable('FFD_START',buf,status=st)
+        if (st.eq.0) read(buf,*) it0
+        call get_environment_variable('FFD_NSTEP',buf,status=st)
+        if (st.eq.0) read(buf,*) nst
+      endif
+      ffdb_on = .not.(it0.lt.0 .or. itime.lt.it0
+     &     .or. itime.ge.it0+nst)
+      end function ffdb_on
+
+      subroutine ffdb_pass_inc()
+!@sum D96: count AFLUX entries within one DYNAM call (= leapfrog pass
+!@+   index 1..5: 1 fwd(MRCH=0), 2 bwd(-1), 3 even(2), 4 odd(-2), 5 even(2)).
+!@+   Reset whenever itime changes.
+      use model_com, only : itime
+      implicit none
+      integer :: ffdb_last,ffdb_np
+      common /ffdbpass/ ffdb_last,ffdb_np
+      save /ffdbpass/
+      integer, save :: init=0
+      if (init.eq.0) then
+        init=1
+        ffdb_last=-1
+        ffdb_np=0
+      endif
+      if (itime.ne.ffdb_last) then
+        ffdb_last=itime
+        ffdb_np=0
+      endif
+      ffdb_np=ffdb_np+1
+      end subroutine ffdb_pass_inc
+
+      subroutine ffdb_open(fu,tag,rt,ph)
+      use model_com, only : itime
+      implicit none
+      integer :: fu
+      character(len=*) :: tag,ph
+      character(len=*) :: rt
+      character(len=200) :: fname
+      integer :: ffdb_last,ffdb_np
+      common /ffdbpass/ ffdb_last,ffdb_np
+      save /ffdbpass/
+      write(fname,'(A,A,I0,A,I0,A,A,A)') 'ffd_',trim(tag),itime,'_p',
+     &     ffdb_np,'_',trim(ph),'.bin'
+      open(fu,file=trim(fname),form='unformatted',access='stream',
+     &     status='replace')
+      end subroutine ffdb_open
+
+      subroutine ffdb_geom(naflux,aftopo,np,ip,jp,md)
+!@sum D96: static geometry/constants for AFLUX/ADVECM/MAtoP/PGF/ADVECV
+!@+   written once (unit 1083, ffd_aflux_geom.bin).  Layout (doubles):
+!@+   scalars[32], then JM-arrays (dxyp,bydxyp,dyp,bydyp,dxp,dxv,dyv,
+!@+   dxyv,dxyn,dxys,ravpn,ravps,rapvn,rapvs,fcor,cosv,imaxj),
+!@+   IM-arrays (siniv,cosiv), LM-arrays (mfix,mfrac), zatmo(IM,JM),
+!@+   ipatch(2,np),jpatch(2,np),md(np).
+      use constant, only : grav,rgas,kapa,bygrav,bykapa,bykapap1,
+     &     bykapap2,kg2mb,sha,radius,omega
+      use resolution, only : im,jm,lm,mtop,mfixs,mfix,mfrac,mdrya
+      use geom, only : dxyp,bydxyp,dyp,bydyp,dxp,dxv,dyv,dxyv,dxyn,
+     &     dxys,ravpn,ravps,rapvn,rapvs,fcor,cosv,imaxj,siniv,cosiv,
+     &     acor,acor2,polwt,dlon,byim
+      use dynamics, only : dt,mincolmass,maxcolmass,do_polefix,nidyn
+      use atm_com, only : zatmo
+      implicit none
+      logical :: aftopo
+      integer :: naflux,np
+      integer :: ip(2,np),jp(2,np),md(np)
+      integer, save :: init=0
+      integer :: fu
+      if (init.ne.0) return
+      init=1
+      fu=1083
+      open(fu,file='ffd_aflux_geom.bin',form='unformatted',
+     &     access='stream',status='replace')
+      write(fu) dble(im),dble(jm),dble(lm),grav,rgas,kapa,bygrav,
+     &     bykapa,bykapap1,bykapap2,kg2mb,mtop,mfixs,mincolmass,
+     &     maxcolmass,dt,dlon,byim,dble(do_polefix),dble(nidyn),
+     &     merge(1d0,0d0,aftopo),dble(np),polwt,acor,acor2,mdrya,sha,
+     &     radius,omega,0d0,0d0,0d0
+      write(fu) dxyp(1:jm)
+      write(fu) bydxyp(1:jm)
+      write(fu) dyp(1:jm)
+      write(fu) bydyp(1:jm)
+      write(fu) dxp(1:jm)
+      write(fu) dxv(1:jm)
+      write(fu) dyv(1:jm)
+      write(fu) dxyv(1:jm)
+      write(fu) dxyn(1:jm)
+      write(fu) dxys(1:jm)
+      write(fu) ravpn(1:jm)
+      write(fu) ravps(1:jm)
+      write(fu) rapvn(1:jm)
+      write(fu) rapvs(1:jm)
+      write(fu) fcor(1:jm)
+      write(fu) cosv(1:jm)
+      write(fu) dble(imaxj(1:jm))
+      write(fu) siniv(1:im)
+      write(fu) cosiv(1:im)
+      write(fu) mfix(1:lm)
+      write(fu) mfrac(1:lm)
+      write(fu) zatmo(1:im,1:jm)
+      write(fu) dble(ip)
+      write(fu) dble(jp)
+      write(fu) dble(md)
+      close(fu)
+      end subroutine ffdb_geom
+
+      subroutine ffdb_aflux_in(ns,im,jm,lm,u,v,ma,masum,me,mesum,
+     &     aftopo,np,ip,jp,md)
+!@sum D96: AFLUX inputs (unit 1081): ffd_aflux_<itime>_p<k>_in.bin =
+!@+   [itime,pass,NS,DT,MRCH], U,V (im,jm,lm), MA (lm,im,jm), MASUM
+!@+   (im,jm), ME (lm,im,jm), MESUM (im,jm).
+      use model_com, only : itime
+      use dynamics, only : dt,mrch
+      implicit none
+      integer :: ns,im,jm,lm,np
+      real*8 :: u(im,jm,lm),v(im,jm,lm),ma(lm,im,jm),masum(im,jm),
+     &     me(lm,im,jm),mesum(im,jm)
+      logical :: aftopo
+      integer :: ip(2,np),jp(2,np),md(np)
+      logical, external :: ffdb_on
+      integer :: fu
+      integer :: ffdb_last,ffdb_np
+      common /ffdbpass/ ffdb_last,ffdb_np
+      save /ffdbpass/
+      if (.not.ffdb_on()) return
+      call ffdb_geom(0,aftopo,np,ip,jp,md)
+      fu=1081
+      call ffdb_open(fu,'aflux_','','in')
+      write(fu) dble(itime),dble(ffdb_np),dble(ns),dt,dble(mrch)
+      write(fu) u
+      write(fu) v
+      write(fu) ma
+      write(fu) masum
+      write(fu) me
+      write(fu) mesum
+      close(fu)
+      end subroutine ffdb_aflux_in
+
+      subroutine ffdb_aflux_out(im,jm,lm,spa0)
+!@sum D96: AFLUX outputs: [itime,pass], MU,MV (im,jm,lm), MW (im,jm,lm-1),
+!@+   CONV, SPA (post AVRX), SPA0 (SPA before AVRX; rows 2..JM-1 only, 0 elsewhere).
+      use model_com, only : itime
+      use dynamics, only : mu,mv,mw,conv,spa
+      implicit none
+      integer :: im,jm,lm
+      real*8 :: spa0(im,jm,lm)
+      logical, external :: ffdb_on
+      integer :: fu
+      integer :: ffdb_last,ffdb_np
+      common /ffdbpass/ ffdb_last,ffdb_np
+      save /ffdbpass/
+      if (.not.ffdb_on()) return
+      fu=1081
+      call ffdb_open(fu,'aflux_','','out')
+      write(fu) dble(itime),dble(ffdb_np)
+      write(fu) mu(1:im,1:jm,1:lm)
+      write(fu) mv(1:im,1:jm,1:lm)
+      write(fu) mw(1:im,1:jm,1:lm-1)
+      write(fu) conv(1:im,1:jm,1:lm)
+      write(fu) spa(1:im,1:jm,1:lm)
+      write(fu) spa0
+      close(fu)
+      end subroutine ffdb_aflux_out
+
+      subroutine ffdb_advecm_in(im,jm,lm,dt1,mold)
+!@sum D96: ADVECM inputs (unit 1082): ffd_aflux_advecm_<itime>_p<k>_in.bin
+!@+   = [itime,pass,DT1,MRCH], MOLD (lm,im,jm), CONV (im,jm,lm), MW (im,jm,lm-1).
+      use model_com, only : itime
+      use dynamics, only : mrch,conv,mw
+      implicit none
+      integer :: im,jm,lm
+      real*8 :: dt1,mold(lm,im,jm)
+      logical, external :: ffdb_on
+      integer :: fu
+      integer :: ffdb_last,ffdb_np
+      common /ffdbpass/ ffdb_last,ffdb_np
+      save /ffdbpass/
+      if (.not.ffdb_on()) return
+      fu=1082
+      call ffdb_open(fu,'aflux_advecm_','','in')
+      write(fu) dble(itime),dble(ffdb_np),dt1,dble(mrch)
+      write(fu) mold
+      write(fu) conv(1:im,1:jm,1:lm)
+      write(fu) mw(1:im,1:jm,1:lm-1)
+      close(fu)
+      end subroutine ffdb_advecm_in
+
+      subroutine ffdb_advecm_out(im,jm,lm,mnew,msum)
+!@sum D96: ADVECM (+MAtoP) outputs: [itime,pass], MNEW (lm,im,jm),
+!@+   MSUM (im,jm), PEDN,PMID,PDSIG,PK (lm,im,jm; layers 1..lm), P (im,jm)
+!@+   as set by MAtoP at the end of ADVECM.
+      use model_com, only : itime
+      use atm_com, only : pedn,pmid,pdsig,pk,p
+      implicit none
+      integer :: im,jm,lm
+      real*8 :: mnew(lm,im,jm),msum(im,jm)
+      logical, external :: ffdb_on
+      integer :: fu
+      integer :: ffdb_last,ffdb_np
+      common /ffdbpass/ ffdb_last,ffdb_np
+      save /ffdbpass/
+      if (.not.ffdb_on()) return
+      fu=1082
+      call ffdb_open(fu,'aflux_advecm_','','out')
+      write(fu) dble(itime),dble(ffdb_np)
+      write(fu) mnew
+      write(fu) msum
+      write(fu) pedn(1:lm,1:im,1:jm)
+      write(fu) pmid(1:lm,1:im,1:jm)
+      write(fu) pdsig(1:lm,1:im,1:jm)
+      write(fu) pk(1:lm,1:im,1:jm)
+      write(fu) p(1:im,1:jm)
+      close(fu)
+      end subroutine ffdb_advecm_out
+
+      subroutine ffdb_pgf_in(im,jm,lm,dt1,mam,ut,vt,mafter,s0,sz)
+!@sum D97: PGF inputs (unit 1084): ffd_pgf_<itime>_p<k>_in.bin =
+!@+   [itime,pass,DT1,MRCH], MAM (lm,im,jm), MAFTER (lm,im,jm),
+!@+   S0,SZ,UT,VT (im,jm,lm), DUT,DVT (module, im,jm,lm).
+      use model_com, only : itime
+      use dynamics, only : mrch,dut,dvt
+      implicit none
+      integer :: im,jm,lm
+      real*8 :: dt1,mam(lm,im,jm),mafter(lm,im,jm),s0(im,jm,lm),
+     &     sz(im,jm,lm),ut(im,jm,lm),vt(im,jm,lm)
+      logical, external :: ffdb_on
+      integer :: fu
+      integer :: ffdb_last,ffdb_np
+      common /ffdbpass/ ffdb_last,ffdb_np
+      save /ffdbpass/
+      if (.not.ffdb_on()) return
+      fu=1084
+      call ffdb_open(fu,'pgf_','','in')
+      write(fu) dble(itime),dble(ffdb_np),dt1,dble(mrch)
+      write(fu) mam
+      write(fu) mafter
+      write(fu) s0
+      write(fu) sz
+      write(fu) ut
+      write(fu) vt
+      write(fu) dut(1:im,1:jm,1:lm)
+      write(fu) dvt(1:im,1:jm,1:lm)
+      close(fu)
+      end subroutine ffdb_pgf_in
+
+      subroutine ffdb_pgf_out(im,jm,lm,ut,vt,pgfu,pgfu0)
+!@sum D97: PGF outputs: [itime,pass], UT,VT (out), DUT,DVT (module,
+!@+   accumulated), GZ, PHI, SPA (=AdM scratch), PGFU0 (before AVRX,
+!@+   0 for rows not set), PGFU (after AVRX); all (im,jm,lm).
+      use model_com, only : itime
+      use dynamics, only : dut,dvt,spa
+      use atm_com, only : gz,phi
+      implicit none
+      integer :: im,jm,lm
+      real*8 :: ut(im,jm,lm),vt(im,jm,lm),pgfu(im,jm,lm),
+     &     pgfu0(im,jm,lm)
+      logical, external :: ffdb_on
+      integer :: fu
+      integer :: ffdb_last,ffdb_np
+      common /ffdbpass/ ffdb_last,ffdb_np
+      save /ffdbpass/
+      if (.not.ffdb_on()) return
+      fu=1084
+      call ffdb_open(fu,'pgf_','','out')
+      write(fu) dble(itime),dble(ffdb_np)
+      write(fu) ut
+      write(fu) vt
+      write(fu) dut(1:im,1:jm,1:lm)
+      write(fu) dvt(1:im,1:jm,1:lm)
+      write(fu) gz(1:im,1:jm,1:lm)
+      write(fu) phi(1:im,1:jm,1:lm)
+      write(fu) spa(1:im,1:jm,1:lm)
+      write(fu) pgfu0
+      write(fu) pgfu
+      close(fu)
+      end subroutine ffdb_pgf_out
+
+      subroutine ffdb_advecv_in(im,jm,lm,dt1,u,v,mmean,mbefor,ut,vt,
+     &     mafter)
+!@sum D98: ADVECV inputs (unit 1085): ffd_advecv_<itime>_p<k>_in.bin =
+!@+   [itime,pass,DT1,MRCH], U,V (im,jm,lm), MMEAN,MBEFOR,MAFTER
+!@+   (lm,im,jm), UT,VT (im,jm,lm), then module MU,MV (im,jm,lm),
+!@+   MW (im,jm,lm-1), SPA (im,jm,lm) as set by this pass's AFLUX.
+      use model_com, only : itime
+      use dynamics, only : mrch,mu,mv,mw,spa
+      implicit none
+      integer :: im,jm,lm
+      real*8 :: dt1,u(im,jm,lm),v(im,jm,lm),mmean(lm,im,jm),
+     &     mbefor(lm,im,jm),mafter(lm,im,jm),ut(im,jm,lm),vt(im,jm,lm)
+      logical, external :: ffdb_on
+      integer :: fu
+      integer :: ffdb_last,ffdb_np
+      common /ffdbpass/ ffdb_last,ffdb_np
+      save /ffdbpass/
+      if (.not.ffdb_on()) return
+      fu=1085
+      call ffdb_open(fu,'advecv_','','in')
+      write(fu) dble(itime),dble(ffdb_np),dt1,dble(mrch)
+      write(fu) u
+      write(fu) v
+      write(fu) mmean
+      write(fu) mbefor
+      write(fu) mafter
+      write(fu) ut
+      write(fu) vt
+      write(fu) mu(1:im,1:jm,1:lm)
+      write(fu) mv(1:im,1:jm,1:lm)
+      write(fu) mw(1:im,1:jm,1:lm-1)
+      write(fu) spa(1:im,1:jm,1:lm)
+      close(fu)
+      end subroutine ffdb_advecv_in
+
+      subroutine ffdb_advecv_out(im,jm,lm,ut,vt)
+!@sum D98: ADVECV outputs: [itime,pass], UT,VT (im,jm,lm).
+      use model_com, only : itime
+      implicit none
+      integer :: im,jm,lm
+      real*8 :: ut(im,jm,lm),vt(im,jm,lm)
+      logical, external :: ffdb_on
+      integer :: fu
+      integer :: ffdb_last,ffdb_np
+      common /ffdbpass/ ffdb_last,ffdb_np
+      save /ffdbpass/
+      if (.not.ffdb_on()) return
+      fu=1085
+      call ffdb_open(fu,'advecv_','','out')
+      write(fu) dble(itime),dble(ffdb_np)
+      write(fu) ut
+      write(fu) vt
+      close(fu)
+      end subroutine ffdb_advecv_out
+
       SUBROUTINE CHECKT (SUBR)
 !@sum  CHECKT Checks arrays for NaN/INF and reasonablness
 !@auth Original Development Team
